# SICP Chapter 2

In [1]:
;; from book -> not technically correct as will return negative
(define (gcd a b)
  (if (= b 0)
      a
      (gcd b (remainder a b))))

;; construct rational number reduced to lowest terms
(define (make-rat n d) 
  (let ((g (gcd n d)))
    (cons (/ n g)
          (/ d g))))

(define (numer x) (car x))
(define (denom x) (cdr x))

(define (add-rat x y)
  (make-rat (+ (* (numer x) (denom y))
               (* (numer y) (denom x)))
            (* (denom x) (denom y))))

(define (sub-rat x y)
  (make-rat (- (* (numer x) (denom y))
               (* (numer y) (denom x)))
            (* (denom x) (denom y))))

(define (mul-rat x y)
  (make-rat (* (numer x) (numer y))
            (* (denom x) (denom y))))

(define (div-rat x y)
  (make-rat (* (numer x) (denom y))
            (* (denom x) (numer y))))

(define (equal-rat? x y)
  (= (* (numer x) (denom y))
     (* (numer y) (denom x))))

(define (print-rat x)
  (newline)
  (display (numer x))
  (display "/")
  (display (denom x)))

In [2]:
(define half (make-rat 1 2))
(print-rat half)


1/2

In [3]:
(define fifth (make-rat 1 5))
(print-rat fifth)


1/5

In [4]:
(define third (make-rat 1 3))
(print-rat third)


1/3

In [5]:
(print-rat (mul-rat half fifth))


1/10

In [6]:
(print-rat (add-rat half fifth))


7/10

In [7]:
(print-rat (add-rat half half))


1/1

In [8]:
(print-rat (add-rat third third))


2/3

## 2.1

In [9]:
(define (make-rat n d) 
  (let ((g (abs (gcd n d)))) ; abs to correct for negative return from gcd
    (if (< d 0)
        (cons (/ (* n -1) g) 
              (/ (* d -1) g))
        (cons (/ n g) 
              (/ d g)))))
    

In [10]:
(print-rat (make-rat 1 2))


1/2

In [11]:
(print-rat (make-rat -1 2))


-1/2

In [12]:
(print-rat (make-rat 1 -2))


-1/2

## 2.2

In [13]:
(define (make-point x y)
  (cons x y))
(define (x-point point)
  (car point))
(define (y-point point)
  (cdr point))

(define (make-segment start-point end-point)
  (cons start-point end-point))
(define (start-segment segment)
  (car segment))
(define (end-segment segment)
  (cdr segment))

(define (midpoint-segment segment)
  (define (avg x y) (/ (+ x y) 2.0))
  (let ((start (start-segment segment)) (end (end-segment segment)))
    (make-point
     (avg (x-point start) (x-point end))
     (avg (y-point start) (y-point end)))))

(define (print-point p)
  (newline)
  (display "(")
  (display (x-point p))
  (display ",")
  (display (y-point p))
  (display ")"))

In [14]:
(print-point (make-point 1 2))


(1,2)

In [15]:
(define start (make-point 1 2))
(define end (make-point 2 4))

(define seg (make-segment start end))
(print-point (midpoint-segment seg))


(1.5,3.0)

## 2.3

In [16]:
;; opposite corners representation
(define (make-rect bottom-left top-right)
  (cons bottom-left top-right))
(define (rect-width rect)
  (abs (- (x-point (car rect)) (x-point (cdr rect)))))
(define (rect-height rect)
  (abs (- (y-point (car rect)) (y-point (cdr rect)))))

(define (rect-perim rect)
  (* 2 (+ (rect-width rect) (rect-height rect))))

(define (rect-area rect)
  (* (rect-width rect) (rect-height rect)))

In [17]:
(define rect (make-rect (make-point 1 2) (make-point 3 5)))

(rect-width rect)

2

In [18]:
(rect-height rect)

3

In [19]:
(rect-perim rect)

10

In [20]:
(rect-area rect)

6

In [21]:
;; bottom left corner + width/height representation
(define (make-rect bottom-left width height)
  (cons bottom-left (cons width height)))
(define (rect-width rect)
  (car (cdr rect)))
(define (rect-height rect)
  (cdr (cdr rect)))

In [22]:
(define rect (make-rect (make-point 1 2) 2 3))

In [23]:
(rect-perim rect)

10

In [66]:
(rect-area rect)

6

## 2.4

In [32]:
;; returns function that takes 2 arguments and applies it to x y
(define (cons x y)
  (lambda (m) (m x y)))

;; z takes a 2 argument function that is passed p q in order
;; returning first arg p implements car function
(define (car z)
  (z (lambda (p q) p)))

;; z takes a 2 argument function that is passed p q in order
;; returning sencond arg q produces cdr
(define (cdr z)
  (z (lambda (p q) q)))

In [27]:
(car (cons 1 2))

1

In [33]:
(cdr (cons 1 2))

2

Applicative-order evaluation to verify `(car (cons x y))` yields `x` for any objects `x` and `y`:
```scheme
(car (cons (x y)))
(car (lambda (m) (m x y)))
((lambda (m) (m x y)) (lambda (p q) p))
((lambda (p q) p) x y)
x
```

## 2.5

Represent pair of nonnegative integers $a, b$ as the integer that is the product $2^a3^b$

2 and 3 are prime -> $2^a3^b$ will give unique values for each all combinations of $a, b$ due to the [Fundamental Theorem of Arithmetic](http://mathworld.wolfram.com/FundamentalTheoremofArithmetic.html)

$a$ can be calculated from the product by the number of times it is evenly divisible by 2.

$b$ can be calculated from the product by the the number of times it is evenly divisible by 3.

Example:

$a=2\ ,b=3$

$2^a3^b=2^2\cdot3^3=108$

Calculating $a$:

$\frac{108}{2}=54$

$\frac{54}{2}=27$

$\frac{27}{2}=13.5$

2 even divisions by 2:

$a=2$

Calculating $b$:

$\frac{108}{3}=36$

$\frac{36}{3}=12$

$\frac{12}{3}=4$

$\frac{4}{3} = 1.3333333$

3 even divisions by 3:

$b=3$




In [11]:
(define (cons a b)
  (* (expt 2 a)
     (expt 3 b)))

; number of times n is evenly divisible by d
(define (num-div n d)
  (define (iter x count)
    (if (= 0 (remainder x d))
        (iter (/ x d) (+ 1 count))
        count))
  (iter n 0))

(define (car x)
  (num-div x 2))

(define (cdr x)
  (num-div x 3))

In [12]:
(car (cons 2 3))

2

In [13]:
(cdr (cons 2 3))

3

## 2.6

In [17]:
(define zero (lambda (f) (lambda (x) x)))

(define (add-1 n)
  (lambda (f) (lambda (x) (f ((n f) x)))))

Substitution to find definition of `one`:
```scheme
(add-1 zero)
(lambda (f) (lambda (x) (f ((zero f) x))))
(lambda (f) (lambda (x) (f ((lambda (x) x) x))))
(lambda (f) (lambda (x) (f x)))
```

In [19]:
;; applies given procedure once
(define one (lambda (f) (lambda (x) (f x))))

Substitution to find definition of `two`:
```scheme
(add-1 one)
(lambda (f) (lambda (x) (f ((one f) x))))
(lambda (f) (lambda (x) (f ((lambda (x) (f x)) x))))
(lambda (f) (lambda (x) (f (f x))))
```

In [20]:
;; applies given procedure twice
(define two (lambda (f) (lambda (x) (f (f x)))))

In [24]:
;; 'wrap' n in m function calls
(define (church+ m n)
  (lambda (f) (lambda (x) ((m f) ((n f) x)))))

In [38]:
;; easy to see number of function applications with increment procedure
(define (increment n)
  (+ n 1))

In [39]:
((zero increment) 0)

0

In [40]:
((zero increment) 1)

1

In [41]:
((zero increment) 2)

2

In [42]:
((one increment) 0)

1

In [43]:
((one increment) 1)

2

In [44]:
((one increment) 2)

3

In [45]:
((two increment) 0)

2

In [46]:
((two increment) 1)

3

In [47]:
((two increment) 2)

4

In [ ]:
(define three (church+ one two))

((three increment) 0)

## 2.7

In [1]:
;; procedures given in question description
(define (make-interval a b) (cons a b))

(define (add-interval x y)
  (make-interval (+ (lower-bound x)
                    (lower-bound y))
                 (+ (upper-bound x)
                    (upper-bound y))))

;; find minimum and maximum of products of the bounds
(define (mul-interval x y)
  (let ((p1 (* (lower-bound x)
               (lower-bound y)))
        (p2 (* (lower-bound x)
               (upper-bound y)))
        (p3 (* (upper-bound x)
               (lower-bound y)))
        (p4 (* (upper-bound x)
               (upper-bound y))))
    (make-interval (min p1 p2 p3 p4)
                   (max p1 p2 p3 p4))))
(define (div-interval x y)
  (mul-interval x
                (make-interval
                 (/ 1.0 (upper-bound y))
                 (/ 1.0 (lower-bound y)))))

In [2]:
(define (lower-bound x) (min (car x) (cdr x)))
(define (upper-bound x) (max (car x) (cdr x)))

In [3]:
;; test with a=lower, b=upper
(define x (make-interval 10 20))

(lower-bound x)

10

In [4]:
(upper-bound x)

20

In [5]:
;; test with a=upper, b=lower
(define y (make-interval 20 10))

(lower-bound y)

10

In [6]:
(upper-bound y)

20

## 2.8
Difference between intervals `x, y`:
- Minimum = lower bound `x` subtract upper bound `y` 
- Maximum = upper bound `x` subtract lower bound `y`

In [7]:
(define (sub-interval x y)
  (make-interval 
   (- (lower-bound x) (upper-bound y))
   (- (upper-bound x) (lower-bound y))))

(define x (make-interval 10 20))
(define y (make-interval 1 5))
(sub-interval x y)

(5 . 19)

## 2.9

`add-interval`:

$[a,b]+[c,d]=[(a+c), (b+d)]$

`width`:

$width([a,b])=\frac{b-a}{2}$

`width` of `add-interval`:

$width([a,b]+[c,d])=width([(a+c), (b+d)])$

$=\frac{(b+d)-(a+c)}{2}$

sum of `width` of two intervals:

$width([a,b]) + width([c,d])=\frac{b-a}{2}+\frac{d-c}{2}$

$=\frac{(b-a)+(d-c)}{2}$

$=\frac{(b+d)-(a+c)}{2}$

`width` of `add-interval` is equal to the sum of `with` of the same two intervals. Therefore the width of the the sum of two intervals is a function only of the widthds of the intervals being added.

In [8]:
(define (width x)
  (/ (- (upper-bound x) (lower-bound x)) 2))

(define a (make-interval 10 20))
(define b (make-interval 30 40))
(define c (make-interval 1 3))

In [9]:
(width a)

5

In [10]:
(width b)

5

In [11]:
(width c)

1

In [12]:
(width (add-interval a b))

10

Intervals `a` and `b` have the same width. Multiplying each of them by `c` gives produce intervals with different widths. Therefore the product of two intervals has a width that is not a function solely of the widths of said intervals.

In [13]:
(width (mul-interval a c))

25

In [14]:
(width (mul-interval b c))

45

## 2.10

In [15]:
(define (zero-span? x)
  (and (<= (lower-bound x) 0)
       (>= (upper-bound x) 0)))

;; throws an error if divide by 0
(define (div-interval x y)
  (if (zero-span? y)
      (error "div-interval" "Cannot divide by interval with span 0")
      (mul-interval x
                (make-interval
                 (/ 1.0 (upper-bound y))
                 (/ 1.0 (lower-bound y))))))

(div-interval (make-interval 10 20) (make-interval -1 1))


Traceback (most recent call last):
  File "In [15]", line 14, col 1, in 'div-interval'
  File "In [15]", line 8, col 7, in 'error'
  File "In [15]", line 8, col 7
RunTimeError: Error in 'div-interval': Cannot divide by interval with span 0



## 2.11
`(mul-interval x y)` produces an interval where the lower bound is the minimum product of `x` and `y`'s bounds  and the upper bound is the maximum product of `x` and `y`'s bounds.

Upper and lower bounds of an interval can be both positive, both negative or negative lower with positive upper i.e. `[+,+]`, `[-,-]` or `[-,+]`
- Can't have positive lower and negative upper as the lower bound would be less than the upper bound

For `mul-interval` this gives the combinations:
```
[+,+] * [+,+]
[+,+] * [-,-]
[+,+] * [-,+]

[-,-] * [-,-]
[-,-] * [+,+]
[-,-] * [-,+]

[-,+] * [-,+]
[-,+] * [+,+]
[-,+] * [-,-]
```

More than 2 multiplications are required for the case `[-,+] * [-,+]` as the product of the two negative lower bounds could be larger than that of the upper bounds.

In all other cases only two multiplications are required as there is only one combination of products that can produce the correct result
- e.g. `[+,+] * [+,+]` -> multiply two upper bounds and multiply two lower bounds.

In [16]:
(define (pos? x) (>= x 0))
(define (neg? x ) (<= x 0))

(define (mul-interval x y)
  (let ((lx (lower-bound x))
        (ux (upper-bound x))
        (ly (lower-bound y))
        (uy (upper-bound y)))
    (cond ((and (pos? lx)
                (pos? ux)
                (pos? ly)
                (pos? uy))
           ;; [+, +] * [+, +]
           (make-interval (* lx ly) (* ux uy)))
          ((and (pos? lx)
                (pos? ux)
                (neg? ly)
                (pos? uy))
           ;; [+, +] * [-, +]
           (make-interval (* ux lx) (* ux uy)))
          ((and (pos? lx)
                (pos? ux)
                (neg? ly)
                (neg? uy))
           ;; [+, +] * [-, -]
           (make-interval (* ux lx) (* lx uy)))
          ((and (neg? lx)
                (pos? ux)
                (pos? ly)
                (pos? uy))
           ;; [-, +] * [+, +]
           (make-interval (* lx uy) (* ux uy)))
          ((and (neg? lx)
                (pos? ux)
                (neg? ly)
                (pos? uy))
           ;; [-, +] * [-, +]
           (make-interval (min (* ux ly) (* lx uy))
                          (max (* lx ly) (* ux uy))))
          ((and (neg? lx)
                (pos? ux)
                (neg? ly)
                (neg? uy))
           ;; [-, +] * [-, -]
           (make-interval (* ux ly) (* lx ly)))
          ((and (neg? lx)
                (neg? ux)
                (pos? ly)
                (pos? uy))
           ;; [-, -] * [+, +]
           (make-interval (* lx ux) (* ux ly)))
          ((and (neg? lx)
                (neg? ux)
                (neg? ly)
                (pos? uy))
           ;; [-, -] * [-, +]
           (make-interval (* lx ux) (* lx ly)))
          ((and (neg? lx)
                (neg? ux)
                (neg? ly)
                (neg? uy))
           ;; [-, -] * [-, -]
           (make-interval (* ux uy) (* lx ly))))))

In [17]:
(define a (make-interval 1 2))
(define b (make-interval -1 2))
(define c (make-interval -1 -2))

In [18]:
(mul-interval a a)

(1 . 4)

In [19]:
(mul-interval a b)

(2 . 4)

In [20]:
(mul-interval a c)

(2 . -1)

In [21]:
(mul-interval b a)

(-2 . 4)

In [22]:
(mul-interval b b)

(-2 . 4)

In [23]:
(mul-interval b c)

(-4 . 2)

In [24]:
(mul-interval c a)

(2 . -1)

In [25]:
(mul-interval c b)

(2 . 2)

In [65]:
(mul-interval c c)

(1 . 4)

## 2.12

In [27]:
;; definitions given in question
(define (make-center-width c w)
  (make-interval (- c w) (+ c w)))

(define (center i)
  (/ (+ (lower-bound i)
        (upper-bound i))
     2))

(define (width i)
  (/ (- (upper-bound i)
        (lower-bound i))
     2))

In [62]:
;; using make-interval directly
(define (make-center-percent c tolerance)
  (let ((unc (* (/ c 100) tolerance)))
    (make-interval (- c unc) (+ c unc))))

;; using provided make-center-width
(define (make-center-percent c tolerance)
  (make-center-width c (* (/ c 100) tolerance)))

(define (percent i)
  (* (/ (width i) (center i)) 100.0))

In [63]:
(define i (make-center-percent 10 25))
i

(15/2 . 25/2)

In [64]:
(center i)

10

In [65]:
(percent i)

25.0

## 2.13

Interval multiplication:

$[a,b]\cdot[c,d] = [\min(ac,ad,bc,bd), \max(ac,ad,bc,bd)]$

Assuming all values are positive:

$[a,b]\cdot[c,d] = [ac, bd]$

Interval represented by its center $c_i$ and tolerance percentage $p_i$ :

$i = [c_i-\frac{c_i\cdot p_i}{100},c_i+\frac{c_i\cdot p_i}{100}]$

$= [c_i\frac{1-p_i}{100},c_i\frac{1+p_i}{100}]$

Multiplying two intervals $i,j$:

$ij=[c_i\frac{1-p_i}{100},c_i\frac{1+p_i}{100}]\cdot[c_j\frac{1-p_j}{100},c_j\frac{1+p_j}{100}]$

$=[c_i\frac{1-p_i}{100}\cdot c_j\frac{1-p_j}{100}, c_i\frac{1+p_i}{100}\cdot c_j\frac{1+p_j}{100}]$

$=[c_ic_j(1-\frac{p_i+p_j}{100}+\frac{p_ip_j}{10000}), c_ic_j(1+\frac{p_i+p_j}{100}+\frac{p_ip_j}{10000})]$

Percentage tolerances are small ->$\frac{p_ip_j}{10000}$ terms will be small -> approximate as:

$ij=[c_ic_j(1-\frac{p_i+p_j}{100}), c_ic_j(1+\frac{p_i+p_j}{100})]$

Percentage tolerance of the product = $p_i+p_j$
- **sum of the tolerances of the two intervals**

In [67]:
;; test with small tolerances
(define i (make-center-percent 100 2))
(define j (make-center-percent 80 1))

(define ij (mul-interval i j))
;; approximately 2+1=3
(percent ij)

2.9994001199760048

## 2.14

In [68]:
;; procedures from question
(define (par1 r1 r2)
  (div-interval
   (mul-interval r1 r2)
   (add-interval r1 r2)))

(define (par2 r1 r2)
  (let ((one (make-interval 1 1)))
    (div-interval
     one
     (add-interval 
      (div-interval one r1)
      (div-interval one r2)))))

In [85]:
;; demonstrate different results between par1 and par2
(define A (make-center-percent 100 3))
(define B (make-center-percent 200 2))

(par1 A B)

(61.928338762214985 . 71.71331058020478)

In [86]:
(par2 A B)

(64.88737201365187 . 68.44299674267101)

In [87]:
;; investigate behaviour with arithmetic expressions
(define A/A (div-interval A A))
(define A/B (div-interval A B))

(center A/A)

1.0018016214593133

In [88]:
(percent A/A)

5.994604855629938

In [89]:
(center A/B)

0.500500200080032

In [90]:
(percent A/B)

4.997001798920647

The expected center of dividing an interval by itself is 1, however in the investigation above, `(center A/A)` returns `1.0018016214593133` - an approximation to 1.